File: frontend/src/lib/components/engine/installer.ts
import { ComponentDefinition, CommandContext, HookExecutionResult } from '../types'
import { DependencyResolver } from './dependency-resolver'
import { CommandExecutor } from './command-executor'

export interface InstallationProgress {
  currentComponent: string
  totalComponents: number
  currentIndex: number
  status: 'pending' | 'installing' | 'completed' | 'error'
  message?: string
}

export type ProgressCallback = (progress: InstallationProgress) => void
export type HookCallback = (hookResult: HookExecutionResult) => Promise<void>

export interface ComponentInstaller {
  install(
    componentIds: string[],
    context: CommandContext,
    onProgress?: ProgressCallback,
    onHook?: HookCallback
  ): Promise<{ success: boolean; errors: string[] }>

  uninstall(
    componentIds: string[],
    context: CommandContext,
    onProgress?: ProgressCallback
  ): Promise<{ success: boolean; errors: string[] }>

  enable(
    componentId: string,
    context: CommandContext
  ): Promise<{ success: boolean; error?: string }>

  disable(
    componentId: string,
    context: CommandContext
  ): Promise<{ success: boolean; error?: string }>
}

export function createComponentInstaller(
  getComponent: (id: string) => ComponentDefinition | undefined,
  getAllComponents: () => Map<string, ComponentDefinition>,
  executeCommand: (cmd: string) => Promise<boolean>
): ComponentInstaller {
  const resolver = new DependencyResolver(getAllComponents())
  const executor = new CommandExecutor(executeCommand)

  return {
    async install(
      componentIds: string[],
      context: CommandContext,
      onProgress?: ProgressCallback,
      onHook?: HookCallback
    ): Promise<{ success: boolean; errors: string[] }> {
      const errors: string[] = []

      let resolved: string[]
      try {
        resolved = resolver.resolve(componentIds)
      } catch (error) {
        return {
          success: false,
          errors: [(error as Error).message],
        }
      }

      for (let i = 0; i < resolved.length; i++) {
        const componentId = resolved[i]
        const component = getComponent(componentId)

        if (!component) {
          errors.push(`Component not found: ${componentId}`)
          continue
        }

        const componentContext = {
          ...context,
          isInstalled: context.componentsStatus[componentId] || false,
        }

        onProgress?.({
          currentComponent: component.name,
          totalComponents: resolved.length,
          currentIndex: i,
          status: 'installing',
          message: `Installing ${component.name}...`,
        })

        try {
          if (component.preInstall) {
            const hookResult = await component.preInstall.execute(componentContext)
            if (hookResult && onHook) {
              await onHook(hookResult)
            }
          }

          const commands = component.install(componentContext)
          const success = await executor.execute(commands)

          if (!success) {
            throw new Error(`Installation command failed for ${component.name}`)
          }

          if (component.postInstall) {
            const hookResult = await component.postInstall.execute(componentContext)
            if (hookResult && onHook) {
              await onHook(hookResult)
            }
          }

          onProgress?.({
            currentComponent: component.name,
            totalComponents: resolved.length,
            currentIndex: i,
            status: 'completed',
            message: `${component.name} installed successfully`,
          })
        } catch (error) {
          const errorMsg = `Failed to install ${component.name}: ${(error as Error).message}`
          errors.push(errorMsg)

          onProgress?.({
            currentComponent: component.name,
            totalComponents: resolved.length,
            currentIndex: i,
            status: 'error',
            message: errorMsg,
          })
        }
      }

      return {
        success: errors.length === 0,
        errors,
      }
    },

    async uninstall(
      componentIds: string[],
      context: CommandContext,
      onProgress?: ProgressCallback
    ): Promise<{ success: boolean; errors: string[] }> {
      const errors: string[] = []

      for (let i = 0; i < componentIds.length; i++) {
        const componentId = componentIds[i]
        const component = getComponent(componentId)

        if (!component) {
          errors.push(`Component not found: ${componentId}`)
          continue
        }

        if (!component.uninstall) {
          errors.push(`${component.name} does not support uninstall`)
          continue
        }

        const componentContext = {
          ...context,
          isInstalled: context.componentsStatus[componentId] || false,
        }

        onProgress?.({
          currentComponent: component.name,
          totalComponents: componentIds.length,
          currentIndex: i,
          status: 'installing',
          message: `Uninstalling ${component.name}...`,
        })

        try {
          if (component.preUninstall) {
            await component.preUninstall.execute(componentContext)
          }

          const commands = component.uninstall(componentContext)
          const success = await executor.execute(commands)

          if (!success) {
            throw new Error(`Uninstall command failed for ${component.name}`)
          }

          if (component.postUninstall) {
            await component.postUninstall.execute(componentContext)
          }

          onProgress?.({
            currentComponent: component.name,
            totalComponents: componentIds.length,
            currentIndex: i,
            status: 'completed',
            message: `${component.name} uninstalled successfully`,
          })
        } catch (error) {
          const errorMsg = `Failed to uninstall ${component.name}: ${(error as Error).message}`
          errors.push(errorMsg)

          onProgress?.({
            currentComponent: component.name,
            totalComponents: componentIds.length,
            currentIndex: i,
            status: 'error',
            message: errorMsg,
          })
        }
      }

      return {
        success: errors.length === 0,
        errors,
      }
    },

    async enable(
      componentId: string,
      context: CommandContext
    ): Promise<{ success: boolean; error?: string }> {
      const component = getComponent(componentId)

      if (!component) {
        return {
          success: false,
          error: `Component not found: ${componentId}`,
        }
      }

      if (!component.enable) {
        return {
          success: false,
          error: `${component.name} does not support enable`,
        }
      }

      const componentContext = {
        ...context,
        isInstalled: context.componentsStatus[componentId] || false,
      }

      try {
        const commands = component.enable(componentContext)
        const success = await executor.execute(commands)

        if (!success) {
          throw new Error(`Enable command failed for ${component.name}`)
        }

        return { success: true }
      } catch (error) {
        return {
          success: false,
          error: `Failed to enable ${component.name}: ${(error as Error).message}`,
        }
      }
    },

    async disable(
      componentId: string,
      context: CommandContext
    ): Promise<{ success: boolean; error?: string }> {
      const component = getComponent(componentId)

      if (!component) {
        return {
          success: false,
          error: `Component not found: ${componentId}`,
        }
      }

      if (!component.disable) {
        return {
          success: false,
          error: `${component.name} does not support disable`,
        }
      }

      const componentContext = {
        ...context,
        isInstalled: context.componentsStatus[componentId] || false,
      }

      try {
        const commands = component.disable(componentContext)
        const success = await executor.execute(commands)

        if (!success) {
          throw new Error(`Disable command failed for ${component.name}`)
        }

        return { success: true }
      } catch (error) {
        return {
          success: false,
          error: `Failed to disable ${component.name}: ${(error as Error).message}`,
        }
      }
    },
  }
}

File: internal/installer/installer.go
package installer

import (
	"fmt"
	"strings"

	"reManager/internal/component"
	"reManager/internal/executor"
	"reManager/internal/vellum"
)

type InstallResult struct {
	Success  bool
	Errors   []string
	DNSError bool
}

type HookCallback func(result *component.HookExecutionResult) error

type Installer struct {
	vellum   *vellum.Client
	metadata *vellum.MetadataStore
	executor executor.CommandExecutor
}

func NewInstaller(v *vellum.Client, meta *vellum.MetadataStore, exec executor.CommandExecutor) *Installer {
	return &Installer{
		vellum:   v,
		metadata: meta,
		executor: exec,
	}
}

func (i *Installer) Install(
	packageNames []string,
	ctx component.CommandContext,
	onProgress executor.ProgressCallback,
	onHook HookCallback,
) InstallResult {
	var errors []string
	var dnsError bool

	fmt.Printf("[DEBUG] Install starting for packages: %v\n", packageNames)

	for idx, pkgName := range packageNames {
		pkg := i.metadata.GetPackage(pkgName)
		displayName := pkgName
		if pkg != nil {
			displayName = pkg.Name
		}

		fmt.Printf("[DEBUG] Installing package %d/%d: %s\n", idx+1, len(packageNames), pkgName)

		if onProgress != nil {
			onProgress(executor.ProgressInfo{
				CurrentComponent: displayName,
				TotalComponents:  len(packageNames),
				CurrentIndex:     idx,
				Status:           executor.StatusInstalling,
				Message:          fmt.Sprintf("Installing %s...", displayName),
			})
		}

		var lastOutput string
		err := i.vellum.AddStreaming(func(line string) {
			lastOutput = line
			fmt.Printf("[DEBUG] vellum output: %s\n", line)
			if strings.Contains(line, "DNS:") {
				dnsError = true
			}
			if onProgress != nil {
				onProgress(executor.ProgressInfo{
					CurrentComponent: displayName,
					TotalComponents:  len(packageNames),
					CurrentIndex:     idx,
					Status:           executor.StatusInstalling,
					Message:          line,
				})
			}
		}, pkgName)

		if err != nil {
			errMsg := fmt.Sprintf("Installation failed for %s: %v (output: %s)", displayName, err, lastOutput)
			fmt.Printf("[DEBUG] Install error: %s\n", errMsg)
			errors = append(errors, errMsg)
			reportError(onProgress, displayName, len(packageNames), idx, errMsg)
			continue
		}
		fmt.Printf("[DEBUG] Package %s installed successfully\n", pkgName)

		hooks := i.metadata.GetHooks(pkgName)
		if hooks != nil && hooks.PostInstall != "" {
			hookFunc := vellum.GetHook(hooks.PostInstall)
			if hookFunc != nil {
				hookResult, err := hookFunc(ctx)
				if err != nil {
					errors = append(errors, fmt.Sprintf("Post-install hook failed for %s: %v", displayName, err))
					reportError(onProgress, displayName, len(packageNames), idx, errors[len(errors)-1])
					continue
				}
				if hookResult != nil && onHook != nil {
					if err := onHook(hookResult); err != nil {
						errors = append(errors, fmt.Sprintf("Hook callback failed for %s: %v", displayName, err))
						reportError(onProgress, displayName, len(packageNames), idx, errors[len(errors)-1])
						continue
					}
				}
			}
		}

		if onProgress != nil {
			onProgress(executor.ProgressInfo{
				CurrentComponent: displayName,
				TotalComponents:  len(packageNames),
				CurrentIndex:     idx,
				Status:           executor.StatusCompleted,
				Message:          fmt.Sprintf("%s installed successfully", displayName),
			})
		}
	}

	return InstallResult{
		Success:  len(errors) == 0,
		Errors:   errors,
		DNSError: dnsError,
	}
}

func (i *Installer) Uninstall(
	packageNames []string,
	ctx component.CommandContext,
	onProgress executor.ProgressCallback,
	onHook HookCallback,
) InstallResult {
	var errors []string

	for idx, pkgName := range packageNames {
		pkg := i.metadata.GetPackage(pkgName)
		displayName := pkgName
		if pkg != nil {
			displayName = pkg.Name
		}

		if onProgress != nil {
			onProgress(executor.ProgressInfo{
				CurrentComponent: displayName,
				TotalComponents:  len(packageNames),
				CurrentIndex:     idx,
				Status:           executor.StatusInstalling,
				Message:          fmt.Sprintf("Uninstalling %s...", displayName),
			})
		}

		hooks := i.metadata.GetHooks(pkgName)
		if hooks != nil && hooks.PreUninstall != "" {
			hookFunc := vellum.GetHook(hooks.PreUninstall)
			if hookFunc != nil {
				hookResult, err := hookFunc(ctx)
				if err != nil {
					errors = append(errors, fmt.Sprintf("Pre-uninstall hook failed for %s: %v", displayName, err))
					reportError(onProgress, displayName, len(packageNames), idx, errors[len(errors)-1])
					continue
				}
				if hookResult != nil && onHook != nil {
					if err := onHook(hookResult); err != nil {
						errors = append(errors, fmt.Sprintf("Hook callback failed for %s: %v", displayName, err))
						reportError(onProgress, displayName, len(packageNames), idx, errors[len(errors)-1])
						continue
					}
				}
			}
		}

		err := i.vellum.DelStreaming(func(line string) {
			if onProgress != nil {
				onProgress(executor.ProgressInfo{
					CurrentComponent: displayName,
					TotalComponents:  len(packageNames),
					CurrentIndex:     idx,
					Status:           executor.StatusInstalling,
					Message:          line,
				})
			}
		}, pkgName)

		if err != nil {
			errMsg := fmt.Sprintf("Uninstall failed for %s: %v", displayName, err)
			errors = append(errors, errMsg)
			reportError(onProgress, displayName, len(packageNames), idx, errMsg)
			continue
		}

		if hooks != nil && hooks.PostUninstall != "" {
			hookFunc := vellum.GetHook(hooks.PostUninstall)
			if hookFunc != nil {
				hookResult, err := hookFunc(ctx)
				if err != nil {
					errors = append(errors, fmt.Sprintf("Post-uninstall hook failed for %s: %v", displayName, err))
				}
				if hookResult != nil && onHook != nil {
					if err := onHook(hookResult); err != nil {
						errors = append(errors, fmt.Sprintf("Hook callback failed for %s: %v", displayName, err))
					}
				}
			}
		}

		if onProgress != nil {
			onProgress(executor.ProgressInfo{
				CurrentComponent: displayName,
				TotalComponents:  len(packageNames),
				CurrentIndex:     idx,
				Status:           executor.StatusCompleted,
				Message:          fmt.Sprintf("%s uninstalled successfully", displayName),
			})
		}
	}

	return InstallResult{
		Success: len(errors) == 0,
		Errors:  errors,
	}
}

func reportError(onProgress executor.ProgressCallback, name string, total, idx int, msg string) {
	if onProgress != nil {
		onProgress(executor.ProgressInfo{
			CurrentComponent: name,
			TotalComponents:  total,
			CurrentIndex:     idx,
			Status:           executor.StatusError,
			Message:          msg,
		})
	}
}
